support building for and from ARM

**Is your feature request related to a problem? Please describe.**
<!--A clear and concise description of what the problem is-->
Currently none of the build commands (like `bazel build //build:test-targets`) work when run from ARM based machines (M1 Macs). This not only makes it impossible to build images but also makes cert-manager development a bit tough.

**Describe the solution you'd like**
<!--A clear and concise description of what you want to happen.-->
Support ARM architecture natively. If not possible then a workaround with Rosetta 2 might be nice too. Shifting from `Bazel` to `Make` might be a step in this direction.

**Describe alternatives you've considered**
<!--A clear and concise description of any alternative solutions or features you've considered.-->
I tried running the terminal with Rosetta 2 and trying the build commands but that failed too.

**Additional context**
<!--Add any other context about the feature request here.-->
See slack [discussion](https://kubernetes.slack.com/archives/CDEQJ0Q8M/p1616499006040200) 



/kind feature
cc @irbekrm

Hi @irbekrm!

1. I'm running `bazel 4.0.0-homebrew`. That is the [latest release](https://github.com/bazelbuild/bazel/releases/tag/4.0.0) and was released on 21st Jan but the [link you mentioned](https://github.com/bazelbuild/bazel/commit/8c7e11a81c35563d951062fe6f00e4e85b3009fb) is of 27th Jan so I'm not sure if that PR is included or not. I will try the instructions mentioned in [this comment](https://github.com/bazelbuild/bazel/pull/12900#issuecomment-769900027) and get back to you.
2. I cloned the repo and ran that command and it worked :) This is the output I got:
```
bazel build //:gazelle_local                                       ✔ 
Starting local Bazel server and connecting to it...
INFO: SHA256 (https://golang.org/dl/?mode=json&include=all) = 64a61556e291c1695f7597037faec67fd48d87feced12541fff23f3a8400cf72
INFO: Analyzed target //:gazelle_local (76 packages loaded, 7696 targets configured).
INFO: Found 1 target...
Target //:gazelle_local up-to-date:
  bazel-bin/gazelle_local_/gazelle_local
INFO: Elapsed time: 73.751s, Critical Path: 13.38s
INFO: 95 processes: 11 internal, 84 darwin-sandbox.
INFO: Build completed successfully, 95 total actions
```

This is weird though. Since I don't have that commit of 27 Jan I'm not entirely sure why this worked 🤔 

3. `name -a` output:
```
Darwin Arshs-MacBook-Air.local 20.3.0 Darwin Kernel Version 20.3.0: Thu Jan 21 00:06:51 PST 2021; root:xnu-7195.81.3~1/RELEASE_ARM64_T8101 arm64
```
4. Here are the error logs from the [Slack conversation](https://kubernetes.slack.com/archives/CDEQJ0Q8M/p1616499006040200):
```
ERROR: /Users/arsh/Projects/Forks/cert-manager/build/BUILD.bazel:92:10: Configurable attribute "srcs" doesn't match this configuration (would a default condition help?).
Conditions checked:
 @io_bazel_rules_go//go/platform:linux_amd64
 @io_bazel_rules_go//go/platform:darwin_amd64
ERROR: Analysis of target '//build:test-targets' failed; build aborted: /Users/arsh/Projects/Forks/cert-manager/build/BUILD.bazel:92:10: Configurable attribute "srcs" doesn't match this configuration (would a default condition help?).
Conditions checked:
 @io_bazel_rules_go//go/platform:linux_amd64
 @io_bazel_rules_go//go/platform:darwin_amd64
FAILED: Build did NOT complete successfully (294 packages loaded, 1893 targets configured)
FAILED: Build did NOT complete successfully (294 packages loaded, 1893 targets configured)
    Fetching @com_github_onsi_ginkgo; Restarting.
    Fetching @bazel_gazelle_go_repository_config; Restarting.
    Fetching @local_config_cc_toolchains; fetching
    Fetching @bazel_gazelle_go_repository_cache; Restarting.
make: *** [verify] Error 1
```
5. I can run any tests you need to perform on my machine so please feel free to let me know 😄 

Thank you so much for working on this!

--- a/build/platforms.bzl
+++ b/build/platforms.bzl
@@ -46,7 +46,7 @@ CLIENT_PLATFORMS = {
 
 TEST_PLATFORMS = {
     "linux": ["amd64"],
-    "darwin": ["amd64"],
+    "darwin": ["amd64", "arm64"],
 }
 
 # Helper which produces the ALL_PLATFORMS dictionary, currently composed of

--- a/hack/bin/BUILD.bazel
+++ b/hack/bin/BUILD.bazel
@@ -18,6 +18,7 @@ genrule(
     name = "com_coreos_etcd",
     srcs = select({
         ":darwin": ["@kubebuilder-tools_darwin_amd64//:etcd"],
+        ":darwin_arm": ["@kubebuilder-tools_darwin_amd64//:etcd"],
         ":k8": ["@kubebuilder-tools_linux_amd64//:etcd"],
     }),
     outs = ["etcd"],
@@ -29,6 +30,7 @@ genrule(
     name = "co_honnef_go_tools_staticcheck",
     srcs = select({
         ":darwin": ["@co_honnef_go_tools_staticcheck_osx//:file"],
+        ":darwin_arm": ["@co_honnef_go_tools_staticcheck_osx//:file"],
         ":k8": ["@co_honnef_go_tools_staticcheck_linux//:file"],
     }),
     outs = ["staticcheck"],
@@ -40,6 +42,7 @@ genrule(
     name = "io_kubernetes_kube-apiserver",
     srcs = select({
         ":darwin": ["@kubebuilder-tools_darwin_amd64//:kube-apiserver"],
+        ":darwin_arm": ["@kubebuilder-tools_darwin_amd64//:kube-apiserver"],
         ":k8": ["@kubebuilder-tools_linux_amd64//:kube-apiserver"],
     }),
     outs = ["kube-apiserver"],
@@ -64,6 +67,8 @@ genrule(
     name = "fetch_oc3",
     srcs = select({
         ":k8": ["@oc_3_11_linux//:file"],
+        ":darwin": ["@oc_3_11_mac//:file"],
+        ":darwin_arm": ["@oc_3_11_mac//:file"],
     }),
     outs = ["oc3"],
     cmd = "cp $(SRCS) $@",

--- a/hack/bin/deps.bzl
+++ b/hack/bin/deps.bzl
@@ -54,7 +54,7 @@ filegroup(
 
     http_archive(
         name = "co_honnef_go_tools_staticcheck_osx",
-        sha256 = "03b100561e3bc14db0b3b4004b102a00cb0197938d23cc40193f269f7b246d2d",
+        sha256 = "7fb41768b8e68aaad397f666d7d5eb9c31abcc4180b5cb6fa7d091cef987eb77",
         urls = ["https://github.com/dominikh/go-tools/releases/download/2021.1/staticcheck_darwin_amd64.tar.gz"],
         build_file_content = """
 filegroup(
@@ -204,6 +204,21 @@ filegroup(
      name = "file",
      srcs = [
         "openshift-origin-client-tools-v3.11.0-0cbc58b-linux-64bit/oc",
+     ],
+     visibility = ["//visibility:public"],
+)
+    """,
+    )
+    http_archive(
+        name = "oc_3_11_mac",
+        sha256 = "75d58500aec1a2cee9473dfa826c81199669dbc0f49806e31a13626b5e4cfcf0",
+        urls = ["https://github.com/openshift/origin/releases/download/v3.11.0/openshift-origin-client-tools-v3.11.0-0cbc58b-mac.zip"],
+        build_file_content =
+         """
+filegroup(
+     name = "file",
+     srcs = [
+        "oc",
      ],
      visibility = ["//visibility:public"],
 )

